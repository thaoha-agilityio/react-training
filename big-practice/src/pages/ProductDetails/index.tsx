import { useCallback, useState } from 'react';
import { useParams } from 'react-router-dom';
import { Box, Spinner } from '@chakra-ui/react';

// Components
import CardDetails from '@components/CardDetails';
import Container from '@components/Container';

// Hooks
import { useCustomToast, useFetchProductDetail } from '@hooks';

// Stores
import { useCartStore, useProductStore } from '@stores';

// Constants
import { SUCCESS_MESSAGES } from '@constants';

// types
import { STATUSES } from '@types';

const ProductDetails = () => {
  // Get the 'uuid' parameter from the url
  const { uuid = '' } = useParams();

  // Get function function from custom toast hook
  const { showToast } = useCustomToast();

  // Fetch product details and check if data is being fetched
  const { isFetching } = useFetchProductDetail(uuid);

  // Get carts from carts store
  const { cart, setCart } = useCartStore();

  const product = useProductStore((state) => state.product);

  const [count, setCount] = useState<number>(1);

  const handleChangeQuantity = useCallback((qty: number) => {
    setCount(qty);
  }, []);

  // Handle add product to cart
  const handleAddToCart = useCallback(() => {
    // Check if the product with the given 'uuid' already exists in the cart
    const existedProductIndex = cart?.findIndex((cart) => cart.id === uuid);

    if (existedProductIndex !== -1) {
      const newCart = [...cart];
      newCart[existedProductIndex].quantity += count;

      setCart(newCart);
    } else {
      setCart([...cart, { ...product, quantity: count }]);
    }

    showToast(STATUSES.SUCCESS, SUCCESS_MESSAGES.ADD_TO_CART);
  }, [cart, count, setCart, uuid]);

  return (
    <>
      <Box bg='yellow.150' w='full' h='100px'></Box>
      <Container>
        {isFetching ? (
          <Box textAlign='center' pt='20px'>
            <Spinner variant='secondary' />
          </Box>
        ) : (
          <CardDetails
            card={product}
            onAddToCart={handleAddToCart}
            initialCount={count}
            onQuantityChange={handleChangeQuantity}
          />
        )}
      </Container>
    </>
  );
};

export default ProductDetails;
