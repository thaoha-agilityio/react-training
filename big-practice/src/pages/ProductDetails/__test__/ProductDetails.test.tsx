import { renderWithRouterAndQuery } from '@helpers';
import ProductDetails from '..';

// Hooks
import * as hooks from '@hooks';
import { MOCK_PRODUCTS } from '@constants';

// Mock hooks
jest.mock('@hooks', () => ({
  __esModule: true,
  ...jest.requireActual('@hooks'),
}));

describe('ProductDetails component', () => {
  it('should render component correctly', () => {
    const { container } = renderWithRouterAndQuery(<ProductDetails />);

    expect(container).toBeInTheDocument();
  });

  it('should render product detail correctly', async () => {
    (jest.spyOn(hooks, 'useFetchProductDetail') as jest.Mock).mockImplementation(() => ({
      data: MOCK_PRODUCTS[0],
      isFetching: false,
    }));

    const { getByText } = renderWithRouterAndQuery(<ProductDetails />);
    const addToCartBtn = getByText('Add To Cart');

    expect(addToCartBtn).toBeInTheDocument();
  });
});
