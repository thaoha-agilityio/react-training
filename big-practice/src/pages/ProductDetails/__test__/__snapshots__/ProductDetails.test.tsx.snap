// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductDetails component should render component correctly 1`] = `
<div>
  <div
    class="css-s39g0e"
  />
  <div
    class="chakra-container css-0"
  >
    <div
      class="css-1mvdat4"
    >
      <div
        class="css-1w4kxfq"
      >
        <div
          class="chakra-stack css-pxhuyg"
        >
          <img
            alt="card-image"
            class="chakra-image__placeholder css-0"
            src="https://www.charlotteathleticclub.com/assets/camaleon_cms/image-not-found-4a963b95bf081c3ea02923dceaeb3f8085e1a654fc54840aac61a57a60903fef.png"
          />
        </div>
        <div
          class="chakra-stack css-15oiwx7"
        >
          <p
            class="chakra-text css-10iahqc"
          >
            name 3
          </p>
          <p
            class="chakra-text css-14h5pyi"
          >
            Rs. 
            15
          </p>
          <p
            class="chakra-text css-kb9q25"
          >
            description
          </p>
          <div
            class="css-2kr1vg"
          >
            <div
              class="css-ddi599"
            >
              <button
                class="chakra-button css-4xx2wk"
                type="button"
              >
                -
              </button>
              <p
                class="chakra-text css-1il0tvo"
              >
                1
              </p>
              <button
                class="chakra-button css-4xx2wk"
                type="button"
              >
                +
              </button>
            </div>
            <button
              class="chakra-button css-4xx2wk"
              type="button"
            >
              Add To Cart
            </button>
          </div>
          <div
            class="chakra-stack css-19bjeu2"
          >
            <div
              class="css-kl5uk3"
            >
              <div
                class="css-gw1u79"
              >
                <p
                  class="chakra-text css-0"
                >
                  SKU
                </p>
              </div>
              <p
                class="chakra-text css-0"
              >
                : SS001
              </p>
            </div>
            <div
              class="css-kl5uk3"
            >
              <div
                class="css-gw1u79"
              >
                <p
                  class="chakra-text css-0"
                >
                  Category
                </p>
              </div>
              <p
                class="chakra-text css-0"
              >
                : Furniture
              </p>
            </div>
            <div
              class="css-kl5uk3"
            >
              <div
                class="css-gw1u79"
              >
                <p
                  class="chakra-text css-0"
                >
                  Tags
                </p>
              </div>
              <p
                class="chakra-text css-0"
              >
                : Sofa, Chair
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
